### ISSUE TITLE
Extensibility of FacetedSearchComponent

### ISSUE THREAD

**Issue by selo-ecube** [commercetools/commercetools-sunrise-java#560]
In our project, due to requirements and refactoring, we want to use FacetedSearchComponent with other page content than ProductOverviewPageContent.

[FacetedSearchComponent.onPageDataReady(PageData)](https://github.com/commercetools/commercetools-sunrise-java/blob/51ddadc840f1c51eb46ab3f82581ac291d70bbff/product-catalog/app/com/commercetools/sunrise/productcatalog/productoverview/search/facetedsearch/FacetedSearchComponent.java#L67) is the method that is affected by this:
```
    @Override
    public void onPageDataReady(final PageData pageData) {
        if (pageData.getContent() instanceof ProductOverviewPageContent) {
            final ProductOverviewPageContent content = (ProductOverviewPageContent) pageData.getContent();
            content.setFacets(createFacetSelectorList(facetBeans));
        }
    }
```

I see three ways of achieving our goal:

1. We copy FacetedSearchComponent and make the changes we require, which is not viable in the long term because we'd need to watch out for bug-fixes and updates in future versions of the original class.
1. You could introduce an interface, e.g. called 'WithFacets', that ProductOverviewPageContent could implement so it could be used in FacetedSearchComponent to remove the close dependency on ProductOverviewPageContent. Our page content classes would then implement the new interface.
1. The reason why we cannot replace FacetedSearchComponent.onPageDataReady(...) with a modified method is that facetBeans is private, as are all other fields and utility methods. Making the utility methods and the non-injected private fields protected would make subclassing useful, but introduce another kind of interface to subclasses, though.

Since the last two options depend on your development policies, I thought it's better to raise this issue for discussion than creating a pull-request.

Thanks a lot for helping out!

**Comment by lauraluiz** [comment]
@selo-ecube we definitely have to decouple all facets from the ProductOverviewPageContent and improve it in many ways. Actually under the label `search` you can see all related topics that we want to have in mind once we start to rework all search classes this year. The search classes were done before the framework conversion thus they still need to be adapted.

Anyway, focusing on this component I like idea 2). Even as a global idea to help many other components: to provide via interface information about the type of objects inside the content. I would rename it to `WithFacetsContent` or something that contains a common suffix for `PageContent` (opposite to controller's interfaces).

Only thing that worries me is how to enable different facet configurations for different pages, wouldn't that be a valid requirement? We can handle that in the future though.

**Comment by selo-ecube** [comment]
Maybe `WithFacetsModel` is a good choice for 2), because `FacetSelectorListBean`
is a `ViewModel`?

But you're right: 2) - alone - would be too inflexible. I don't see right now that we'd like to use different facet configurations for different pages. We currently aim at using category-specific facet configurations and will use a custom FacetSearchSelectorListFactory. But right now we put a feature toggle into our copy of FacetedSearchComponent.onPageDataReady() to switch off facet display until development is complete. Such a change would only be possible with option 3).

**Comment by lauraluiz** [comment]
`WithFacets(View)Model` it is then. I like it better too.

Regarding viability of option 3), I think that units like controller components which have such specific use cases behind, should be `final` classes that can be configured via `application.conf` and making them more generic, like with the interfaces you proposed. That's the path I consider more maintainable for the future. If the use case changes too much, then the solution should be creating a new controller component, not overriding it.

So maybe we could create and abstract class that handles the this use case for any facets and then the concrete one with the facets coming from the configuration.

**Comment by selo-ecube** [comment]
Yes, that's a valid approach towards 3) and extracting an abstract base class of `FacetSearchSelectorListFactory` sounds like a good idea, thank you!

**Comment by mmoelli** [comment]
@selo-ecube Can we close this issue then or is more work needed here?

**Comment by selo-ecube** [comment]
I used option 1) in our local project to proceed. The remainder of the comments was about how to solve the issue in the Sunrise framework, where @lauraluiz is in charge.

**Comment by lauraluiz** [comment]
@mmoelli we need to keep this open, as I'm already applying the solution to the current working branch but it's still not merged :)

**Comment by lauraluiz** [comment]
Fixed in version > v1.0.0-M1
See https://github.com/commercetools/commercetools-sunrise-java/blob/v1.0.0-M1.1/common/app/com/commercetools/sunrise/search/facetedsearch/AbstractFacetedSearchSelectorControllerComponent.java
Thanks @selo-ecube for the ideas!
